File: IT/1/Funzioni/Esercizi/C++/esercizio26.cpp
/* file esercizio26.cpp */

#include <iostream>
#include <cmath>
#include <vector>

using std::cout;
using std::endl;
using std::vector;

void pFact(int n, vector<int>& fact);

int main() {
	for (int i = 4; i < 20; i++) {
		vector<int> fact;
		pFact(i, fact);
		cout << i;
		if (fact.size() > 1) {
			cout << " = ";
			for (int j = 0; j < fact.size(); j++) {
				cout << fact[j];
				if (j < fact.size()-1) cout << " * ";
			}
			cout << endl;
		}
		else {
			cout << " primo." << endl;
		}
	}
}

void pFact(int n, vector<int>& fact) {
	if (n <= 3) return;
	int m = n;
	double sr = sqrt(n);
	for (int f = 2; f < (int)sr+1; f++) {
		while (m % f == 0) {
			fact.push_back(f);
			m = m / f;
		}
	}
	if (m > 1) fact.push_back(m);
}

File: IT/1/Funzioni/Esercizi/Python/esercizio26.py
# file esercizio26.py

from __future__ import print_function
from math import sqrt
from sys import stdout
from os import linesep

def pFact(n, f, fact):
	if i <= 3: return
	m = n
	sr = sqrt(n)
	for f in range(2, int(sr)+1):
		while m % f == 0:
			fact.append(f)
			m = m / f
	if (m > 1): fact.append(m)

for i in range(4, 20):
	fact = []
	pFact(i, 2, fact)
	print("{} = ".format(i), end="")
	l = len(fact)
	if l > 1:
		for j in range(l):
			p = fact[j]
			print(p, end="")
			if j < l-1: print(" * ", end="")
			else: stdout.write(linesep)
	else:
		print("primo.")

